Simplify RID Model (#260)
dotnet/designs: accepted/2022/simplify-rid-model.md

Note: RIDs are almost entirely oriented around calling native code. However, the domain is broader, more generally oriented on native interop and calling [Foreign Function Interfaces (FFI)](https://en.wikipedia.org/wiki/Foreign_function_interface), including matching the environment. For example, we'd still likely need to use RIDs for interacting with another managed platform like Java, possibly with the use of native code or not.

Note: RIDs are a close cousin to Target Framework Monikers (TFMs), however have less polished UX and don't describe the same concept (although they overlap).

Note: This topic area applies to other development platform, and is typically a domain of more challenging UX. Whether .NET is better or worse off than other development platforms for native code targeting is hard to say.

- https://peps.python.org/pep-0600/
- https://peps.python.org/pep-0656/
- https://doc.rust-lang.org/rustc/platform-support.html

## Biggest problems

The following are the biggest problems we see:

> - We get requests to add new RIDs ([Asianux Server 8](https://github.com/dotnet/runtime/issues/2129), [CBL-Mariner], (https://github.com/dotnet/runtime/issues/65566)), with no end in sight.

Reviewer: ```suggestion
- We get requests to add new RIDs ([Asianux Server 8](https://github.com/dotnet/runtime/issues/2129), [CBL-Mariner](https://github.com/dotnet/runtime/issues/65566), [Anolis Linux](https://github.com/dotnet/runtime/pull/66132)), with no end in sight.
```

- Existing RID authoring sometimes [doesn't work with new OS releases](https://github.com/dotnet/runtime/issues/65152).
- Reasoning about [portable vs non-portable Linux RIDs](https://github.com/dotnet/runtime/pull/62942).
- We have to read and process `runtime.json` at application startup, which has a (unmeasured) performance impact.
- The fact that the RID graph is so large (and continuing to grow) demonstrates that we chose a poor design point.

## General approach

- Ensure `runtimes.json` is correct.

> - Freeze `runtimes.json`.

Reviewer 2: Freeze completely? How are we going to support a new OSes (e.g. Fuchsia/Zirkon) ?

Author: They use an existing host, such that the OS uses the `linux-x64` RID. If they have specific needs, then we can discuss it.

The host and `runtimes.json` must remain compatible. The RIDs known by the host must be a subset of `runtimes.json` RIDs. We may need to update `runtimes.json` to ensure that the RIDs known by the host are present in that file.

## Minimum CRT/libc version

This scheme doesn't explicitly define a CRT/libc version. Instead, we will define a minimum CRT/libc version, per major .NET version. That minimum CRT/libc version will form a contract and should be documented, per .NET version. Practically, it defines the oldest OS version that you can run an app on, for a given .NET version.

For .NET 7:

- For Windows, .NET will target the Windows 10 CRT.
- For Linux with glibc, .NET will target CentOS 7.
- For Linux with musl, .NET will target the oldest supported Alpine version.

For .NET 8, we'll continue with the model. However, we will no longer be able to use [CentOS 7 (due to EOL)](https://wiki.centos.org/About/Product), but will need to adopt another distro that provides us with an old enough glibc version such that we can offer broad compatibility.

> For .NET 9, we'll likely drop support for Windows 10 and instead target the Windows 11 CRT.

Reviewer 3: Does this mean that .NET will "likely drop support for Windows 10" almost a full year before Windows 10's end-of-life date (November 2024 vs October 2025)?

Note: It appears that some other folks have been [reacting to CentOS 7 not being a viable compilation target](https://gist.github.com/wagenet/35adca1a032cec2999d47b6c40aa45b1) for much longer.

## Legal RIDs

RIDs are defined by legal combinations of operating system, chip architecture, and C-runtime. We support both triplets (primarily native code) and singles (only processor-agnostic managed code).

A given RID is used to describe:

- This system supports this RID, per the .NET host.
- This code will run on / requires a system that supports this RID, per the author.

The following table describes RIDs supported by the .NET host. It is exhaustive, per the hosts provided by Microsoft.

| RID        | Description of scope |
|------------|----------------------|
| unix       | All Unix-based OSes (macOS and Linux), versions, and architecture builds. |
| win        | All Windows versions and architecture builds. |

> | linux-arm64| All Linux glibc-based distros, for Arm64. |

Reviewer 4: `linux-arm32`? Did you forget it or will it stop being supported? (I don't personally care, just asking)

Author: Good catch. An oversight.

https://github.com/dotnet/designs/pull/260/commits/e7582b563f3359e6889f6ef07d4e3591073ba45a

## Source-build

A major design tenet of the RID system is maximizing portability of code. This is particularly true for Linux. That makes sense from the perspective of Microsoft wanting to make one build of .NET available across many Linux distros, separately for both glibc and musl. It makes less sense for distros themselves building .NET from source and publishing binaries to a distro package feed.

We sometimes refer to `linux-x64` (and equally to `linux-arm64` and `linux-x86`) as the "portable Linux RID". As mentioned in the libc section, this RID establishes a wide glibc compatibility range. In addition, the RID also establishes broad compatibility across distros (and their associated versions) for .NET dependencies, like OpenSSL and ICU. The way this particular form of compatibility shows up is observable but is nuanced (and isn't explained here).

We'll use Red Hat as an example of an organization that builds .NET from source, to complete this discussion. They have raised concerns on [source-build not playing nicely with `linux-x64`](https://github.com/dotnet/runtime/pull/62942#issuecomment-1056942235).

> For Red Hat, it makes sense to accept and support `linux-x64` assets, but not to produce or offer them. Instead, Red Hat would want to produce `rhel-x64` assets. It's easiest to describe this design-point in terms of concrete scenarios.

Reviewer 5: The rid chosen by a source-build maintainer should not be the same for distribution versions that are not binary compatible. That means source-build rids are likely to have versions in them.

Can you use `rhel.8-x64` instead of `rhel-x64` to make this visible?

Author: Yes. I realized this after publishing the doc. I'll update this.

Author: Ughh. I added a mess of an answer. Thoughts?

https://github.com/dotnet/designs/pull/260/commits/840970625f490356d6de38aceea67320987f4fa9

Reviewer 5: lgtm.

> However, for that to work, we would need to add `rhel.8-x64` to `runtimes.json`. Otherwise, NuGet won't be able to realize that `rhel.8-x64` should be treated as `linux-x64` for restoring NuGet assets. However, we want to freeze `runtimes.json`. That topic is left as an unresolved issue in this spec, at least for now.

dotnet/runtime has support for patching the `runtime.json` graph with the source-build rid. See https://github.com/dotnet/runtime/pull/50818 by @ericstj.

Author: Yup. You are right. Will resolve that. Eric separately gave me the same context.

Author: Added to the doc.

**NuGet packages** -- NuGet authors will typically value breadth and efficiency. For example, a package like `SkiaSharp` might target `linux-x64`, but not `rhel-x64`. There is no technical reason for that package to include distro-specific assets. Also, if the author produced a `rhel-x64` asset, they would need to consider producing an `ubuntu-x64` and other similar assets for other distros, and that's not scalable. As a result, it makes sense for Red Hat to support `linux-x64`. There is no technical downside to Red Hat doing so.

> **Runtime and host packs** -- The .NET SDK doesn't include all assets that are required for all scenarios, but instead relies on many assets being available on a NuGet feed (nuget.org or otherwise). That's a good design point for a variety of reasons. Runtime packs are a concrete scenario, which are used for building self-contained apps. Red Hat should be able to (A) build RHEL-specific runtime packages, (B) publish them to a feed of their choice, and (C) enable their users to download them by specifying a RHEL-specific RID as part of a typical .NET CLI workflow. RHEL-specific runtime packs would not be compatible in the same way as `linux-x64`. They would support `linux-x64` NuGet assets, but would only (presumably) support being run on Red Hat Enterprise Linux family OSes. For example, the RHEL-specific runtime for RHEL 8 would only support the default OpenSSL version that is provided in the RHEL 8 package feed.

Reviewer 5: Can you include the RHEL-specific rid explicitly in this section (as it gets used in the rest of the doc)?
Something like:

```suggestion
**Runtime and host packs** -- The .NET SDK doesn't include all assets that are required for all scenarios, but instead relies on many assets being available on a NuGet feed (nuget.org or otherwise). That's a good design point for a variety of reasons. Runtime packs are a concrete scenario, which are used for building self-contained apps. Red Hat should be able to (A) build RHEL-specific runtime packages, (B) publish them to a feed of their choice, and (C) enable their users to download them by specifying a RHEL-specific RID (`rhel-x64`) as part of a typical .NET CLI workflow. RHEL-specific packs would not be compatible in the same way as `linux-x64`. They packs would (presumably) only support being run on Red Hat Enterprise Linux family OSes. For example, the RHEL-specific runtime for RHEL 8 would only support the default OpenSSL version that is provided in the RHEL 8 package feed.
```

Author: Your edit makes it sound like you would not support `linux-x64` NuGet assets. That doesn't match my intuition. Can you elaborate on your thinking?

Reviewer 5: I meant to clarify that the packs built by Red Hat have the `rhel-x64` rid, and not the `linux-x64` rid.

> RHEL-specific packs would not be compatible in the same way as `linux-x64`. They packs would (presumably) only support being run on Red Hat Enterprise Linux family OSes.

Maybe this makes it more clear:

> RHEL-specific packs would not be compatible in the same way as `linux-x64` packs. The `rhel-x64` packs would (presumably) only support being run on Red Hat Enterprise Linux family OSes.

Reviewer 5: Adding some more text:

> RHEL-specific packs would not be compatible in the same way as the `linux-x64` packs (built by Microsoft). The `rhel-x64` packs would (presumably) only support being run on Red Hat Enterprise Linux family OSes, while the `linux-x64` packs work across a range of distros.

Author: You are still removing my text on `linux-x64` NuGet assets. I'm talking about a third-party NuGet package. That remains relevant and needs to be explained.

How about this:

> RHEL-specific packs would not be compatible in the same way as the `linux-x64` packs (built by Microsoft). The `rhel-x64` packs would (presumably) only support being run on Red Hat Enterprise Linux family OSes, while the `linux-x64` packs work across a range of distros. However,`linux-x64` NuGet assets (in third-party NuGet packages) would be usable, compatible, and restorable with both Red Hat (distro-specific) and Microsoft (portable) builds of .NET.

Reviewer 5: > linux-x64 NuGet assets (in third-party NuGet packages) would be usable, compatible, and restorable with both Red Hat (distro-specific) and Microsoft (portable) builds of .NET.

I think this belongs in the previous paragraph on _NuGet packages_.

Author: I'm sorry. You were right the whole time. Thanks for being patient.

Check my latest commit.

Author: https://github.com/dotnet/designs/pull/260/commits/8eb6dae68635228eacfea2dbdc0a430114267ea8

Reviewer 5: lgtm, thanks.

- https://peps.python.org/pep-0600/
- https://peps.python.org/pep-0656/
- https://doc.rust-lang.org/rustc/platform-support.html

## Biggest problems

The following are the biggest problems we see:

- We get requests to add new RIDs ([Asianux Server 8](https://github.com/dotnet/runtime/issues/2129), [CBL-Mariner](https://github.com/dotnet/runtime/issues/65566), [Anolis Linux](https://github.com/dotnet/runtime/pull/66132)), with no end in sight.
- RID [authoring is needed for new versions of existing OSes](https://github.com/dotnet/runtime/issues/59803) and [sometimes doesn't work when that doesn't happen](https://github.com/dotnet/runtime/issues/65152)
- Reasoning about [portable vs non-portable Linux RIDs](https://github.com/dotnet/runtime/pull/62942).
- We have to read and process `runtime.json` at application startup, which has a (unmeasured) performance impact.
- The fact that the RID graph is so large (and continuing to grow) demonstrates that we chose a poor design point.

## General approach

- Ensure `runtimes.json` is correct.
- Freeze `runtimes.json`.

> - Continue to use `runtimes.json` for all NuGet scenarios.

Reviewer 6: Do you have any pointers to what sort of things are covered by nuget scenarios?

Author: I don't. Someone else might, but it is really two parts:

- Folder names that are using in NuGet packages.
- The algorithm used by `dotnet restore` in matching the host or user-supplied RID to NuGet package folders.

Fair?

This behavior only applies for portable apps. RID-specific apps will not probe for RID-specific assets.

There are other RIDs in `runtimes.json` today. This new scheme will not support those. The `runtimes.json` host compat feature will need to be used to support those, which is likely very uncommon. Even if they are quasi-common, we will likely still make this change and require some combination of the ecosystem to adapt and app developers to opt-in to the `runtimes.json` compat mode.

The host and `runtimes.json` must remain compatible. The RIDs known by the host must be a subset of `runtimes.json` RIDs. We may need to update `runtimes.json` to ensure that the RIDs known by the host are present in that file.

## Minimum CRT/libc version

This scheme doesn't explicitly define a CRT/libc version. Instead, we will define a minimum CRT/libc version, per major .NET version. That minimum CRT/libc version will form a contract and should be documented, per .NET version. Practically, it defines the oldest OS version that you can run an app on, for a given .NET version.

For .NET 7:

- For Windows, .NET will target the Windows 10 CRT.

> - For Linux with glibc, .NET will target CentOS 7.

Reviewer 6: Is that the same for all architectures? For example, in addition to x64, will the CentOS 7 version of glibc also be the baseline for `arm64` and `s390x` architectures?

Author: That would be my base assumption.

Reviewer 5: They can be different because they are anyhow incompatible.

If you don't need to support CentOS 7 on an architecture, there may be build time benefits to using something newer.

For s390x, Microsoft isn't building any `linux-s390` artifacts for .NET, and the Red Hat artifacts will use a source-build rid (like `rhel.8-s390x`).
The glibc baseline serves for NuGet package authors that want build binaries that work across a range of distros. The basline should be the lowest version of CentOS/RHEL that supports s390x.

Author: I added some text on this.

Note: It appears that some other folks have been [reacting to CentOS 7 not being a viable compilation target](https://gist.github.com/wagenet/35adca1a032cec2999d47b6c40aa45b1) for much longer.

## Legal RIDs

RIDs are defined by legal combinations of operating system, chip architecture, and C-runtime. We support both triplets (primarily native code) and singles (only processor-agnostic managed code).

A given RID is used to describe:

- This system supports this RID, per the .NET host.
- This code will run on / requires a system that supports this RID, per the author.

> The following table describes RIDs supported by the .NET host. It is exhaustive, per the hosts provided by Microsoft.

Reviewer 6: How do we deal with architectures like `s390x` that Microsoft doesn't support and/or provide builds for?

Author: That's what this is about: https://github.com/dotnet/designs/pull/260/files#diff-20aa11d3c5c8d54fc1c45f4b81927255d3951fdd7cf58589fc6d03a2dd2a6a1bR221

The proposal is to NOT add any new RIDs, but enable source-builds to add additive `runtimes.json` files that augment the graph. So, frozen is kind a soft term. It means that we want to avoid creating any new official RIDs, but that distro maintainers can create their own for their own distro. The assumption is that they are only needed to create a link between the user and the runtime/host packs and not NuGet. As a result, these RIDs don't need to be stored anywhere else. We'll accept a file like `runtimes-s390x.json` in upstream. It would probably be <20 lines.

Fair?

Reviewer 6: > The assumption is that they are only needed to create a link between the user and the runtime/host packs and not NuGet

In the SkiSharp scenario, the authors of SkiaSharp might want to publish `linux-s390x` assets on nuget.org. What would that look like?

Alternatively, do you think it might be worth it to blanket-add all the current architectures supported by mono to runtime.json before freezing it?

Reviewer 5: > In the SkiSharp scenario, the authors of SkiaSharp might want to publish linux-s390x assets on nuget.org. What would that look like?

They can add an so file in their NuGet package at runtimes/linux-s390x, and then the package will support s390x.

```
$ wget --no-check-certificate https://www.nuget.org/api/v2/package/SkiaSharp.NativeAssets.Linux/2.88.0-preview.232
$ unzip 2.88.0-preview.232
$ cd runtimes/
$ tree
.
├── linux-arm
│   └── native
│       └── libSkiaSharp.so
├── linux-arm64
│   └── native
│       └── libSkiaSharp.so
├── linux-musl-x64
│   └── native
│       └── libSkiaSharp.so
└── linux-x64
    └── native
        └── libSkiaSharp.so
```

The `linux-s390x` rid could be added as part of source-build which means it will be usable on the resulting .NET distribution.

The reason for adding it to upstream `runtimes.json` is to make it a known rid for to **all** .NET distributions. Then they can publish against that rid. Otherwise it would be unknown.

Author: That's a good point. Agree. Is that good @omajid? If so, I'll update with the spec with the wording from @tmds.

Author: Updated the doc.

Date: April 2022

[Runtime IDs (RIDs)](https://docs.microsoft.com/dotnet/core/rid-catalog) are one of the most challenging features of the product. They are necessary, impact many product scenarios, and have very obscure and challenging UX. They also equally affect development and runtime and sometimes prevent .NET from running on operating systems that Microsoft doesn't support. This document proposes a plan to improve some aspects of the way RIDs work, particularly for scenarios that we've found to be insufficient for our needs.

## Context

RIDs are (for the most part) modelled as a graph of [target triplet](https://wiki.osdev.org/Target_Triplet) symbols that describe legal combinations of operating system, chip architecture, and C-runtime, including an extensive fallback scheme. This graph is codified in [`runtime.json`](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json), which is describes the RID catalog. It is a massive (database in a) file. That's a lot of data to reason about, hard to edit correctly, and a bear to maintain.

Examples of triplets today are:

- `win-x86`
- `macos-arm64`
- `linux-arm64`

> - `linux-x64-musl`

Reviewer 7: If this is supposed to be an actual RID (and it feels like it is since the first 3 are RIDs and not triplets), the value is flipped. In .NET RIDs it is: `linux-musl-x64`.

Author: Sorry. I always write that wrong. Thanks!

```suggestion
- `linux-musl-x64`
```

The current system is unnecessarily complex, expensive to maintain, makes it difficult to manage community needs, and costly/fragile at runtime. More simply, it is useful, but overshoots the needs we actually have.

Note: RIDs are almost entirely oriented around calling native code. However, the domain is broader, more generally oriented on native interop and calling [Foreign Function Interfaces (FFI)](https://en.wikipedia.org/wiki/Foreign_function_interface), including matching the environment. For example, we'd still likely need to use RIDs for interacting with another managed platform like Java, possibly with the use of native code or not.

Note: RIDs are a close cousin to Target Framework Monikers (TFMs), however have less polished UX and don't describe the same concept (although they overlap).

Note: This topic area applies to other development platform, and is typically a domain of more challenging UX. Whether .NET is better or worse off than other development platforms for native code targeting is hard to say.

- https://peps.python.org/pep-0600/
- https://peps.python.org/pep-0656/
- https://doc.rust-lang.org/rustc/platform-support.html

## Biggest problems

> The following are the biggest problems we see:

Reviewer 7: I believe https://github.com/dotnet/runtime/issues/59803 should be added to this list. The difference between the first bullet below and this issue is that the first bullet below are for new OS names. https://github.com/dotnet/runtime/issues/59803 is for new versions of OSes that we already know about. (And needs servicing changes as well.)

Author: I added that. Feel free to edit if I messed up the description.

- There are processor-agnostic managed-code scenarios where a RID single is relevant, like Windows-only code (for example, accessing the registry) that works the same x64, Arm64 and x86. The same is true for macOS and Linux.
- The RID that `dotnet --info` returns will always match the first in this hard-coded set of RIDs.

Let's assume that an app is running on Alpine 3.16, the host will probe for the following RIDs:

- `linux-musl-x64`
- `unix`

Ubuntu would be similar:

- `linux-arm64`
- `unix`

macOS would be similar:

- `osx-arm64`
- `unix`

Windows would similar:

- `win-x86`
- `win`

Note: `win7` and `win10` also exist (as processor-specific and -agnostic). Ideally, we don't have to support those in the host-specific scheme, but we need to do some research on that.

Note: A mix of processor-specific RIDs are used above, purely to demonstrate the range of processor-types supported. Each operating system supports a variety of processor-types, and this is codified in their associated RIDs.

More generally, the host will probe for:

- This environment, RID triplet (OS-CRuntime-arch)
- This environment, RID single (OS-only).

> Note: The abstract `unix` RID is used for macOS and Linux, instead of a concrete RID for those OSes.

Reviewer 7: What if I wanted to create a managed .dll that worked across all `macOS` architectures? The same scenario as the above:

> There are processor-agnostic managed-code scenarios where a RID single is relevant, like Windows-only code (for example, accessing the registry) 

Except for `macOS`-specific. Like P/Invoking into macOS native APIs.

The same scenario can be made for Linux and reading certain files like `/proc/self`, etc.

Reviewer 7: Or is the expectation that .NET Libraries should be using TFMs (`net6.0-macos`, `net6.0-windows`, etc) for those scenarios?

Author: My understanding is that the use of macOS-specific code is limited and that we rely primarily on `unix` in our own codebase. As a result, I saw `osx` as something we should wait for more concrete scenarios. If that's wrong, we can add `osx` to the macOS probing list now.

The TFMs are really fully separate. They can be use for the purpose you intend, but not for code that is intended as `net6.0`. For that, you need to use RIDs, as you know.

Author: I'll add this to the spec as something to resolve. It will mostly go away, since `osx-x64` will eventually age out.

This behavior only applies for portable apps. RID-specific apps will not probe for RID-specific assets.

There are other RIDs in `runtimes.json` today. This new scheme will not support those. The `runtimes.json` host compat feature will need to be used to support those, which is likely very uncommon. Even if they are quasi-common, we will likely still make this change and require some combination of the ecosystem to adapt and app developers to opt-in to the `runtimes.json` compat mode.

The host and `runtimes.json` must remain compatible. The RIDs known by the host must be a subset of `runtimes.json` RIDs. We may need to update `runtimes.json` to ensure that the RIDs known by the host are present in that file.

## Minimum CRT/libc version

This scheme doesn't explicitly define a CRT/libc version. Instead, we will define a minimum CRT/libc version, per major .NET version. That minimum CRT/libc version will form a contract and should be documented, per .NET version. Practically, it defines the oldest OS version that you can run an app on, for a given .NET version.

For .NET 7:

- For Windows, .NET will target the Windows 10 CRT.

> - For Linux with glibc, .NET will target CentOS 7.

Reviewer 8: Should this be codified to a concrete glibc version (e.g., 2.17 that is used on CentOS 7)) rather than CentOS version itself to make this clear for other distros?

Author: Yes and no. I think that's somewhat confusing the point. My understanding is that we're targeting the glibc that CentOS 7 has, whatever it is, making the specific version irrelevant. Fair?

Author: A little more context. We target the glibc in CentOS 7 to build our `linux-x64` product. If there are distros with an older glibc, then we don't support them, by definition. If .NET is run on a distro with a same/newer glibc, then it is all good. If you build .NET yourself on a distro with a different glibc version, that doesn't matter. It has nothing to do with this plan.

Reviewer: I agree with @tkapin that this should be speced using the glibc version. A few reasons:
- If I have some random Linux distro with given glibc version, I should be able to tell whether it is compatible without downloading CentOS 7 just to find the glibc version that happens to be in it. We should not take favorites with the distros.
- The reference distro used by Microsoft official build is likely going to change over time.

It is fine for the spec to explain the underlying reason behind the choice (e.g. that it is the version in CentOS that is the oldest among supported mainstream Linux distros currently).

This applies for both glibc and musl.

Reviewer 5: _RHEL/CentOS 7 is so old, there can't be anything older._ :smile: 

It has glibc v2.17.

```
$ podman run centos:7 yum list installed glibc
Loaded plugins: fastestmirror, ovl
Installed Packages
glibc.x86_64                        2.17-317.el7                         @CentOS
```

Reviewer 8: To add one more on top of @jkotas reasons, if we expressed this dependency in terms of distro version, that distro would either have to guarantee that it doesn't offer glibc upgrade to another version within the same version of the distro, or we would have to ensure that we support the earliest version of glibc shipps with that distro version.

We should definitely list the OSes that are supposed to be compatible, but since the primary thing is compatibility with glibc, I think it should be canonically expressed via glibc version with the distro version deduced from it rather than the other way around.

I'm not having an extremely strong opinion about this though so if you feel the former would be more transparent, I'm ok with that.

Author: Fixed. You guys are right.

We sometimes refer to `linux-x64` (and equally to `linux-arm64` and `linux-x86`) as the "portable Linux RID". As mentioned in the libc section, this RID establishes a wide glibc compatibility range. In addition, the RID also establishes broad compatibility across distros (and their associated versions) for .NET dependencies, like OpenSSL and ICU. The way this particular form of compatibility shows up is observable but is nuanced (and isn't explained here).

We'll use Red Hat as an example of an organization that builds .NET from source, to complete this discussion. They have raised concerns on [source-build not playing nicely with `linux-x64`](https://github.com/dotnet/runtime/pull/62942#issuecomment-1056942235).

For Red Hat, it makes sense to accept and support `linux-x64` assets, but not to produce or offer them. Instead, Red Hat would want to produce `rhel-x64` assets. It's easiest to describe this design-point in terms of concrete scenarios.

> **NuGet packages** -- NuGet authors will typically value breadth and efficiency. For example, a package like `SkiaSharp` might target `linux-x64`, but not `rhel-x64`. There is no technical reason for that package to include distro-specific assets. Also, if the author produced a `rhel-x64` asset, they would need to consider producing an `ubuntu-x64` and other similar assets for other distros, and that's not scalable. As a result, it makes sense for Red Hat to support `linux-x64`. There is no technical downside to Red Hat doing so.

Reviewer 5: > As a result, it makes sense for Red Hat to support `linux-x64`. There is no technical downside to Red Hat doing so.

Red Hat doesn't do anything to support `linux-x64`. It is people publishing `linux-x64` assets (like Microsoft) that declare compatibility with a range of glibc based Linux distros (like RHEL).

Author: I think we're talking past one another.

The question is:

- When a dev using the RHEL variant of .NET restores a NuGet asset with a `linux-x64` as an asset (and no other Linux compatible asset), will they get it?

The `linux-x64` pack assets that Microsoft publishes are very different than the ones that the community publishes in NuGet packages.

Reviewer 5: Maybe use RHEL instead of Red Hat.

Using Red Hat makes it sound like we do the work. We don't.
The `SkiaSharp` NuGet authors are the ones doing the work by providing native assets that (should) work across the `linux-x64` distros, including RHEL.

Author: Fixed that.

Note: It appears that some other folks have been [reacting to CentOS 7 not being a viable compilation target](https://gist.github.com/wagenet/35adca1a032cec2999d47b6c40aa45b1) for much longer.

## Legal RIDs

RIDs are defined by legal combinations of operating system, chip architecture, and C-runtime. We support both triplets (primarily native code) and singles (only processor-agnostic managed code).

A given RID is used to describe:

- This system supports this RID, per the .NET host.
- This code will run on / requires a system that supports this RID, per the author.

The following table describes RIDs supported by the .NET host. It is exhaustive, per the hosts provided by Microsoft.

| RID        | Description of scope |
|------------|----------------------|
| unix       | All Unix-based OSes (macOS and Linux), versions, and architecture builds. |
| win        | All Windows versions and architecture builds. |

> | linux-arm32| All Linux glibc-based distros, for Arm32. |

Reviewer 6: This should be [`linux-arm`](https://github.com/TobiasLaving/runtime/blob/312c66f1fc2f749f56612999cb1adab9ca7fde59/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json#L1460), right?

Author: ```suggestion
| linux-arm| All Linux glibc-based distros, for Arm32. |
```

> | linux-arm32| All Linux glibc-based distros, for Arm32. |

Reviewer 6: Are there any plans to add [`linux-armv6`](https://github.com/dotnet/runtime/pull/62594) here?

Author: Ya. I'll add it. We can decide whether it fully makes sense later.

## Context

RIDs are (for the most part) modelled as a graph of [target triplet](https://wiki.osdev.org/Target_Triplet) symbols that describe legal combinations of operating system, chip architecture, and C-runtime, including an extensive fallback scheme. This graph is codified in [`runtime.json`](https://github.com/dotnet/runtime/blob/313982dff3457c8291e8b88cd99785f6920319f0/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json), which we refer to as the RID catalog. It is a massive (database in a) file. That's a lot of data to reason about, hard to edit correctly, and a bear to maintain.

Note: `runtimes.json` is  a generated file, from [runtimeGroups.props](https://github.com/dotnet/runtime/blob/2042eb2a56112539bf9e0b9538972fdd84e381a0/src/libraries/Microsoft.NETCore.Platforms/src/runtimeGroups.props).

Examples of triplets today are:

- `win-x86`

> - `macos-arm64`

Reviewer 9: ```suggestion
- `osx-arm64`
```

# Simplify RID Model

Date: April 2022

[Runtime IDs (RIDs)](https://docs.microsoft.com/dotnet/core/rid-catalog) are one of the most challenging features of the product. They are necessary, impact many product scenarios, and have very obscure and challenging UX. They also equally affect development and runtime and sometimes prevent .NET from running on operating systems that Microsoft doesn't support. This document proposes a plan to improve some aspects of the way RIDs work, particularly for scenarios that we've found to be insufficient for our needs.

## Context

> RIDs are (for the most part) modelled as a graph of [target triplet](https://wiki.osdev.org/Target_Triplet) symbols that describe legal combinations of operating system, chip architecture, and C-runtime, including an extensive fallback scheme. This graph is codified in [`runtime.json`](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json), which is describes the RID catalog. It is a massive (database in a) file. That's a lot of data to reason about, hard to edit correctly, and a bear to maintain.

Reviewer 10: > target triplet

AFAIK, target triplets have more cons than pros. It really depends on target audience, but there are negative connotations with the concept of `triplet` in native tooling communities. They consider it vague and archaic concept. e.g. for the modern ISA, the compiler toolchain [documentation](https://www.sifive.com/blog/all-aboard-part-1-compiler-args) was written in terms of ABI (`-mabi`), architecture (`-march`) and microarchitecture (`-mtune`). In case of .NET tooling, `--arch` and `--os` are  relatable concepts.

Author: `--arch` and `--os` are indeed nicer to work with. They came out of our effort to make it easier to support emulated environments (like Rosetta 2 on macOS). They are however just a convenience wrapper over the RID concept. More importantly, NuGet packages need a currency for describing assets intended for various. The triplets are the simplest scheme we've found for that. We haven't found a better system.

Reviewer 10: I completely understand and it is probably too late to do anything about it. Since RID is inspired by triplets, I just wanted to point out that modern toolchains prefer explicitly specifying each aspect of platform individually over having a triplet scheme to interpret magic strings, which parser/consumer would always need to remember. They also started their journey with triplets but over the years, found it as an unnecessary overhead.

Author: That all makes sense when you are using a compiler to produce artifacts that you will use. I fail to see how that works for a package manager that supports a broad computing ecosystem. Those specific flags also seem too low-level. By definition, they are not intended (per my read) to be used outside of the RISC-V ecosystem.

Reviewer 10: > I fail to see how that works for a package manager that supports a broad computing ecosystem.

`rid: linux-x64` versus `arch: x64, os: linux`

Author: OK. You literally just mean using attributes. Got it. Yes, that would have been a different initial design.

Package folks would then look like:

AttributeSpecificAssets
  A
    attributes.json
    Foo.dll
  B
    attributes.json
    Foo.dylib

I'm sure it wouldn't look exactly like that, but that's the basic idea. It could work. A nice characteristic of the current design is that the folder names (for A and B) encode the target so the lookup mechanism is easier. Other things (as you've called out) with that design are worse.

Reviewer 10: Yup, the parsing looks like this: https://github.com/dotnet/runtime/blob/b42adaddff9047e32c9d7f4145746dbb579758f2/src/libraries/Microsoft.NETCore.Platforms/src/RID.cs#L59 and we do that in msbuild scripts and C/C++ separately. :)

Reviewer: > AFAIK, target triplets have more cons than pros. It really depends on target audience, but there are negative connotations with the concept of triplet in native tooling communities.

I agree that the classic `machine-vendor-operatingsystem` triplets have archaic feel, due to the `vendor` field and to all sorts of cryptic suffixes that tend to be appended to them.

Most RIDs are `operatingsystem-machine` that does not the same feel. Maybe the doc should not lead with explaining RIDs as target triplets to avoid giving readers wrong impression.

Python `Platform Tag` defined in [PEP 0425[(https://peps.python.org/pep-0425/) is much more similar to RIDs than target triplets. The definition of the Python `Platform Tag` does not make any associations with target triplets.

Reviewer 9: Target triplets do indeed have problems and the parsing is annoying. But I think the fundamental problem is the same -- in order to provide binary compatibility between components, we need to have an encoding for whatever's necessary for binary compatiblity on that platform.

For Mac and Windows, OS version and architecture is probably sufficient. For Linux, we'll also need libc type (and version, although it could be implicitly tied to the TFM).

Reviewer: Yes, we are in agreement about what we want to do. This is just a discussion about wording - whether it is helpful to describe RID as target triplet or not.

Reviewer 10: Yes, RID is essentially a close-enough hint and we cannot reliably capture the "desired system configuration" with it accurately. As you pointed out in other thread @agocke, that was not the goal RID was designed to achieve. We perform run time introspections and we have abstraction layers in place to adapt to the system we are running on; in (a fex/countable) places where it makes a difference in dotnet/runtime. This is probably the extent to which 3P package, with OS specific native artifacts,  rely on RID.

Author: I am happy to remove the linkage with target triplet if that is what folks want.

- https://peps.python.org/pep-0600/
- https://peps.python.org/pep-0656/
- https://doc.rust-lang.org/rustc/platform-support.html

## Biggest problems

The following are the biggest problems we see:

- We get requests to add new RIDs ([Asianux Server 8](https://github.com/dotnet/runtime/issues/2129), [CBL-Mariner](https://github.com/dotnet/runtime/issues/65566), [Anolis Linux](https://github.com/dotnet/runtime/pull/66132)), with no end in sight.
- RID [authoring is needed for new versions of existing OSes](https://github.com/dotnet/runtime/issues/59803) and [sometimes doesn't work when that doesn't happen](https://github.com/dotnet/runtime/issues/65152)
- Reasoning about [portable vs non-portable Linux RIDs](https://github.com/dotnet/runtime/pull/62942).
- We have to read and process `runtime.json` at application startup, which has a (unmeasured) performance impact.
- The fact that the RID graph is so large (and continuing to grow) demonstrates that we chose a poor design point.

## General approach

- Ensure `runtimes.json` is correct.
- Freeze `runtimes.json`.

> - New RIDs will only be added for [interchange across multiple platforms](https://github.com/dotnet/designs/pull/260#discussion_r843009872), which isn't expected.

Reviewer 11: When you say

> interchange across multiple platforms

You really mean "between different distributions of .NET", right?

Author: ```suggestion
- New RIDs will only be added for [interchange between different distributions of .NET](https://github.com/dotnet/designs/pull/260#discussion_r843009872), which isn't expected.
```

- Continue to use `runtimes.json` for all NuGet scenarios.
- Disable using `runtimes.json` for all host scenarios, by default (starting with a TBD .NET version).
- Enable a host compatibility mode that uses `runtimes.json` via MSBuild property (which writes to), `runtimeconfig.json`, and a host CLI argument.
- Implement a new algorithmic host scheme for RID selection, enabled by default.

The new scheme will be based on the following:

> - Each host is built with a hard-coded set of RIDs that it probes for, both triplets and singles.

Reviewer 11: Does this mean that we either

(1) Hard code the (large set) fallbacks into the host
(2) Evaluate the fallbacks at build time and put them in the (small) set of RID specific folders the host is built for?

Author: Neither. This is covered later in the doc. Each host will know about 2-3 RIDs and none other. If that doesn't work, there will be a fall-back mechanism (which will just be a MSBuild + `runtimeconfig.json` property) to use the RID catalog. Make sense?

- The RID that `dotnet --info` returns will always match the first in this hard-coded set of RIDs.

Let's assume that an app is running on Alpine 3.16, the host will probe for the following RIDs:

- `linux-musl-x64`
- `unix`

A glibc distro like Debian would be similar:

- `linux-arm64`
- `unix`

macOS would be similar:

- `osx-arm64`
- `unix`

Note: The abstract `unix` RID is used for macOS and Linux, instead of a concrete RID for those OSes.

Note: We may want to add `osx` to describe macOS-specific processor-agnostic managed code. If we find some scenarios that need it, we should add it. 

Windows would similar:

- `win-x86`
- `win`

Note: `win7` and `win10` also exist (as processor-specific and -agnostic). Ideally, we don't have to support those in the host-specific scheme, but we need to do some research on that.

Note: A mix of processor-specific RIDs are used above, purely to demonstrate the range of processor-types supported. Each operating system supports a variety of processor-types, and this is codified in their associated RIDs.

More generally, the host will probe for:

- This environment, RID triplet (OS-CRuntime-arch)

> - This environment, RID single (OS-only).

Reviewer 11: If we build a host for a specific RID, why do we need to do any probing at runtime? Asked differently, why can't we do all the probing at build time?

Author: Because of portable apps.

The host will implement "first one wins" probing logic for selecting a RID-specific asset.

This behavior only applies for portable apps. RID-specific apps will not probe for RID-specific assets.

There are other RIDs in `runtimes.json` today. This new scheme will not support those. The `runtimes.json` host compat feature will need to be used to support those, which is likely very uncommon. Even if they are quasi-common, we will likely still make this change and require some combination of the ecosystem to adapt and app developers to opt-in to the `runtimes.json` compat mode.

The host and `runtimes.json` must remain compatible. The RIDs known by the host must be a subset of `runtimes.json` RIDs. We may need to update `runtimes.json` to ensure that the RIDs known by the host are present in that file.

## Minimum CRT/libc version

> This scheme doesn't explicitly define a CRT/libc version. Instead, we will define a minimum CRT/libc version, per major .NET version. That minimum CRT/libc version will form a contract and should be documented, per .NET version. Practically, it defines the oldest OS version that you can run an app on, for a given .NET version.

Reviewer: Using C runtime version to define the OS version makes sense on Linux. It is the best approximation of the OS version to use for determining binary compatibility (of binaries written in C/C++).

Using C runtime version for this purpose makes less sense on Windows and macOS. Windows and macOS have well-defined OS version. Both Windows (Windows SDK) and macOS (XCode) native toolchains are oriented on using the OS version for this purpose ([WINVER](https://docs.microsoft.com/en-us/cpp/porting/modifying-winver-and-win32-winnt) on Windows and `-mmacos-version-min` on macOS). We should use OS version for determining binary compatibility on Windows and macOS.

Author: Do you want me to add that text, or adapt its meaning into the current text?

Reviewer: I think adapting the meaning into the current text would look better. I would lead with minimum OS version as a documented contract, per .NET version. libc version is Linux-specific approximation of OS version across distros for the purposes of this contract.

We sometimes refer to `linux-x64` (and equally to `linux-arm64` and `linux-x86`) as the "portable Linux RID". As mentioned in the libc section, this RID establishes a wide glibc compatibility range. In addition, the RID also establishes broad compatibility across distros (and their associated versions) for .NET dependencies, like OpenSSL and ICU. The way this particular form of compatibility shows up is observable but is nuanced (and isn't explained here).

We'll use Red Hat as an example of an organization that builds .NET from source, to complete this discussion. They have raised concerns on [source-build not playing nicely with `linux-x64`](https://github.com/dotnet/runtime/pull/62942#issuecomment-1056942235).

For Red Hat, it makes sense to accept and support `linux-x64` NuGet assets, but not to produce or offer them. Instead, Red Hat would want to produce `rhel-x64` runtime pack assets. It's easiest to describe this design-point in terms of concrete scenarios.

> **NuGet packages** -- NuGet authors will typically value breadth and efficiency. For example, a package like `SkiaSharp` might target `linux-x64`, but not `rhel-x64`. There is no technical reason for that package to include distro-specific assets. Also, if the author produced a `rhel-x64` asset, they would need to consider producing an `ubuntu-x64` and other similar assets for other distros, and that's not scalable. The expectation is that the `rhel-x64` .NET supports `linux-x64` NuGet assets, enabling NuGet authors to target a minimal set of RIDs.

Reviewer 12: I feel like this document generally, and this section specifically, is making some assumptions that I don't think are justified. There are absolutely scenarios where distro-specific assets need to be shipped in a NuGet package.

As an example of this, take a look at https://www.nuget.org/packages/LibGit2Sharp.NativeBinaries/2.0.306

LibGit2Sharp is a managed wrapper around the native libgit2 library and has to accommodate the native dependencies of libgit2. One of the dependencies is OpenSSL, which varies wildly among all the distros that .NET supports.

In order to have a reasonable chance of having a native binary that just works on all platforms, I have had to spend a lot of time understanding the RID graph and shipping enough distro-specific binaries to cover all the supported Linux distros

I will point out that libgit2 has recently changed how they are binding to OpenSSL which has let me simplify things down to more of an "ideal" situation (see the [newer package](https://www.nuget.org/packages/LibGit2Sharp.NativeBinaries/2.0.315)), but that was largely out of my control since I'm not a libgit2 maintainer.

If the ability to ship distro-specific assets in a NuGet package and have the proper one selected had not been a feature that I could count on, then I would not have been able to support LibGit2Sharp on Linux at all for the past several years. Maybe I'm misunderstanding something, but it seems like this proposal is removing this feature.

Reviewer 6: Mind you, https://github.com/libgit2/libgit2sharp/pull/1714 changed the model used by libgit2sharp. AFAIK, the change basically makes libgit2sharp try and load all native libgit2 libraries one by one until one of them can be loaded successfully (ie, links to a version of OpenSSL available on the system). The actual name (and RID) is not too relevant in this scenario.

I feel like your use-case would still work if nuget did not "know" about distro-specific RIDs.

Author: This is great feedback. We discussed it.

There are three models:

- Don't support specific distros as a general concept (the current proposal).
- Fully support specific distros as a general concept (some variation on status-quo).
- Adopt a model that is a variation on the Python manylinux wheel plan, including support for OpenSSL.

I like what @omajid is proposing. It's the same approach we use for OpenSSL. We don't use RIDs. Also, the current proposal works for the vast majority of cases and is a massive simplification. I'm hoping we can keep it.

Reviewer 5: libgit2sharp was suffering from the rid scalability issue: it was unable to deal with unknown rids.

Using `NativeLibrary.SetDllImportResolver` it now tries the different `so` files. The code is here: https://github.com/libgit2/libgit2sharp/blob/97bee65fd296f1c7dd2d1d64581c170f45b584e1/LibGit2Sharp/Core/NativeMethods.cs#L99-L121

This way, the library can have its own logic for picking `so` files. It's not limited to/by rids.

Reviewer 12: Yes, this logic was added, but I consider it a fallback solution, and not something desirable to rely on as the primary logic, for a couple of reasons:
- It only works if the entire runtimes folder is shipped as part of the application, which is only going to be true if you use one of the cross-platform publishing options
- Having to attempt to load every binary until you find one that works seems like it would have a negative impact on startup perf.

If libraries no longer can assume that the best native asset is selected for them by the framework/NuGet, then that means they have to manually ensure that all binaries are included in the publish output to be able to select from them at runtime. That can pretty massively increase the size of the output. Using LibGit2Sharp.NativeBinaries 2.0.306 as an example, that means going from having a single ~1MB .so file in your output to having 9 different copies, and that's assuming I'd have some way to know to only copy the Linux binaries and not the entire runtimes folder.

Reviewer 5: > It only works if the entire runtimes folder is shipped as part of the application, which is only going to be true if you use one of the cross-platform publishing options

This way of trimming it isn't scalable as you need to build and include an `so` file for every platform you want to trim.

A scalable solution is to allow to control the trimming more specifically. For example add a property to csproj that enables the user to choose the supported OpenSSL versions, and include `so` files based on that.

> Having to attempt to load every binary until you find one that works seems like it would have a negative impact on startup perf.

I doubt this has a measurable impact.

> Using LibGit2Sharp.NativeBinaries 2.0.306 as an example, that means going from having a single ~1MB .so file in your output to having 9 different copies

It seems the latest version doesn't have distro specific versions. Why did it change?

Author: > perf

.NET probes for OpenSSL. No one has ever submitted a perf report on that.

@bartonjs -- can you point to the probing code for OpenSSL?

Reviewer: > can you point to the probing code for OpenSSL?

OpenSSL probing code in dotnet/runtime: https://github.com/dotnet/runtime/blob/main/src/native/libs/System.Security.Cryptography.Native/opensslshim.c#L54-L121

OpenSSL probing in libgit2: 
https://github.com/libgit2/libgit2/blob/82f526a9129373d899782c411774c8e396cae4e4/deps/ntlmclient/crypt_openssl.c#L63-L69 . Notice that it does not have support for probing for OpenSSL 3 yet so it is going to break once distros switch to OpenSSL 3 only. dotnet/runtime has probing for OpenSSL3 that adds a whole new level of shimming and complexity.

> perf

The probing is relatively expensive when looked at in isolation (it is file I/O). It is relatively cheap compared to e.g. all JITing that happens during startup of a typical .NET app.

> I see there is a linux-x64 asset. What's the intended behavior of that asset? By definition, it can run on any (glibc) distro.

Yes, this works like bow that since libgit2 added a configuration to probe for OpenSSL (https://github.com/libgit2/libgit2/pull/5974) and the linux-x64 binaries in the nuget package are built with this option on.

Reviewer 12: > This way of trimming it isn't scalable as you need to build and include an so file for every platform you want to trim.

That's not how publishing works currently, though. It uses the RID graph to select the best option from the assets in the package. My goal was always to have enough coverage so that the supported .NET platforms would just work. That *was* broken sometimes due to the RID graph not being updated...

> It seems the latest version doesn't have distro specific versions. Why did it change?

I mentioned that earlier here: libgit2 has recently added a dynamic binding option for OpenSSL, which has let me simplify the package.

> I see there is a linux-x64 asset. What's the intended behavior of that asset? By definition, it can run on any (glibc) distro.

The set of binaries in the package is a result of analyzing the distros that .NET Core 3.1 supported and the RID graph to ship the fewest number of binaries with the maximum amount coverage. The binary shipped as `linux-x64` was intended to be a fallback that would work when a more specific binary was not chosen. That meant I needed to understand the glibc and OpenSSL versions in each distro to be able to construct the set of binaries I shipped.

This was a large effort, but the end result was that most/all the distros would work out of the box. However, when the RID graph was not updated for a new .NET Core release and a new distro/version incorrectly defaulted back to `linux-x64`, then that fallback would fail. That's where @tmds PR came in and added an additional fallback to try and see if one of the binaries in the package might work even though the RID graph wasn't able to select it.

Reviewer: > For example add a property to csproj that enables the user to choose the supported OpenSSL versions, and include so files based on that.

The problem does not end with OpenSSL. The same problem exists for many other libraries.

I believe that the only truly scalable option for Linux ecosystem is to have building from source as an option (https://github.com/NuGet/Home/issues/9631). Building from source is capable of producing a binary that works for the specific configuration of your system without the package maintainer pre-building it for you.

Reviewer 12: > The problem does not end with OpenSSL. The same problem exists for many other libraries.

Yes, and this was the main point of the feedback I was trying to raise here. The proposal currently assumes that there is never a reason to care about the specific distro you're running on, but that is demonstrably not true.

This all comes down to there being a disconnect between the .NET ecosystem expectations of sharing pre-compiled binaries, and the Linux ecosystem that expects you to share source and compile it on your own system. Shipping a Linux binary in a package is always going to be an uphill battle, so I would like to see the .NET side of things evolve in a way that makes it easier, not harder.

Reviewer 5: > Notice that it does not have support for probing for OpenSSL 3 yet so it is going to break once distros switch to OpenSSL 3 only. dotnet/runtime has probing for OpenSSL3 that adds a whole new level of shimming and complexity.

Let's assume OpenSSL 3 won't be supported by a libgit2 shimming (immediately) and you want to solve it issue using distro rids.

Fedora 36 should use the OpenSSL 3 build.

Fedora 37 should be using the OpenSSL 3 build also. That doesn't work automagically today, because the rid grah doesn't describe a supports relationship between Fedora 37 and Fedora 36.

For your package to remain maintainable, this needs to be addressed (and preferably not by bloating the graph but by making the rid logic smarter).

I've only mentioned Fedora here, but OpenSSL will be adopted by Ubuntu, Debian, so you need to manage this problem for them too.

Your package now looks like this:

```
linux-x64 -> supports OpenSSL 1, 2, 3 and picks using `NativeLibrary`
fedora.36 -> OpenSSL 3
ubuntu.22.04 -> OpenSSL 3
debian.12 -> OpenSSL 3
...
```

Note that you are including the distro specific assets solely for the purpose of trimming. They bloat your package and require you to track in what version a distro adopts a new OpenSSL version.

> The proposal currently assumes that there is never a reason to care about the specific distro you're running on

`NativeLibrary` can be used, and should be used to make `linux-x64` work across a range of distros.

What is lost, is the ability to trim against a rid.

Reviewer 5: @bording I hope this example helped illustrate how unscalable the rid mechanism is?

And, if it should stay, inheritance between successive distro versions is a must-have, which is now missing.

Reviewer 5: > And, if it should stay, inheritance between successive distro versions is a must-have, which is now missing.

To elaborate. The rids as we have them are not scalable. The alternative to the simple model that is proposed here, is to have _semantic rids_. That means rids follow a `<distro>.<version>-<arch>` naming scheme. Information is derived from that instead of hard coding it in the json file (e.g. Fedora 36 -> Fedora 35). The json file should then only contain non-derivable information (like what distros are glibc/musl based).

This is a much more complex solution.

**Runtime and host packs** -- The .NET SDK doesn't include all assets that are required for all scenarios, but instead relies on many assets being available on a NuGet feed (nuget.org or otherwise). That's a good design point for a variety of reasons. Runtime packs are a concrete scenario, which are used for building self-contained apps. Red Hat should be able to (A) build RHEL-specific runtime packages, (B) publish them to a feed of their choice, and (C) enable their users to download them by specifying a RHEL-specific RID as part of a typical .NET CLI workflow. RHEL-specific runtime packs would only be compatible with Red Hat Enterprise Linux family OSes. For example, the RHEL-specific runtime for RHEL 8 would only support the default OpenSSL version that is provided in the RHEL 8 package feed.

**Host RID support** -- The Red Hat provided host would need to probe for assets that might be included in the app, much like was described earlier. It would use the following scheme, here specific to x64.

- `rhel.8-x64`
- `linux-x64`
- `unix`

> A source-built RID (here `rhel.8-x64`) will typically be distro-specific and versioned. Red Hat would naturally want to build .NET specifically and separately for RHEL versions, like RHEL7 and RHEL 8, since the packages available in each of their versioned Red Hat package feeds will be different. There might be other differences to account for as well. An unversioned RID (like `rhel-x64`) would not enable that.

Reviewer 10: > Red Hat would naturally want to build .NET specifically and separately for RHEL versions,

Do we know if other programming platforms (Go, Python, Node.js etc.) also use RH specific triplets? If they do not (which I think is the case), then why do we see it natural requirement for .NET, specifically?

Reviewer 5: Go is well known for its ability to to produce native executables for different platforms.
It has no dependency on a c-library, instead it has its own Go-implementation that makes syscalls directly.
Consequently it doesn't have the need to describe this dependency as a rid.

I don't know the story for building native executables from Python/JavaScript.
